libobs/audio-monitoring/nix/pulseaudio-wrapper.c

```diff
@@ -0,0 +1,333 @@
+	pa_proplist_sets(p, PA_PROP_APPLICATION_ICON_NAME, "obs");
+	pa_proplist_sets(p, PA_PROP_MEDIA_ROLE, "production");
+
+	return p;
+}
+
+/**
+ * Initialize the pulse audio context with properties and callback
+ */
+static void pulseaudio_init_context()
+{
+	pulseaudio_lock();
+
+	pa_proplist *p = pulseaudio_properties();
+	pulseaudio_context = pa_context_new_with_proplist(
+		pa_threaded_mainloop_get_api(pulseaudio_mainloop),
+		"OBS-Monitor", p);
+
+	pa_context_set_state_callback(pulseaudio_context,
+	                              pulseaudio_context_state_changed, NULL);
+
+	pa_context_connect(pulseaudio_context, NULL, PA_CONTEXT_NOAUTOSPAWN,
+	                   NULL);
+	pa_proplist_free(p);
+
+	pulseaudio_unlock();
+}
+
+/**
+ * wait for context to be ready
+ */
+static int_fast32_t pulseaudio_context_ready()
+{
+	pulseaudio_lock();
+
+	if (!PA_CONTEXT_IS_GOOD(pa_context_get_state(pulseaudio_context))) {
+		pulseaudio_unlock();
+		return -1;
+	}
+
+	while (pa_context_get_state(pulseaudio_context) != PA_CONTEXT_READY)
+		pulseaudio_wait();
+
+	pulseaudio_unlock();
+	return 0;
+}
+
+int_fast32_t pulseaudio_init()
+{
+	pthread_mutex_lock(&pulseaudio_mutex);
+
+	if (pulseaudio_refs == 0) {
+		pulseaudio_mainloop = pa_threaded_mainloop_new();
+		pa_threaded_mainloop_start(pulseaudio_mainloop);
+
+		pulseaudio_init_context();
+	}
+
+	pulseaudio_refs++;
+
+	pthread_mutex_unlock(&pulseaudio_mutex);
+
+	return 0;
+}
+
+void pulseaudio_unref()
+{
+	pthread_mutex_lock(&pulseaudio_mutex);
+
+	if (--pulseaudio_refs == 0) {
+		pulseaudio_lock();
+		if (pulseaudio_context != NULL) {
+			pa_context_disconnect(pulseaudio_context);
+			pa_context_unref(pulseaudio_context);
+			pulseaudio_context = NULL;
+		}
+		pulseaudio_unlock();
+
+		if (pulseaudio_mainloop != NULL) {
+			pa_threaded_mainloop_stop(pulseaudio_mainloop);
+			pa_threaded_mainloop_free(pulseaudio_mainloop);
+			pulseaudio_mainloop = NULL;
+		}
+	}
+
+	pthread_mutex_unlock(&pulseaudio_mutex);
+}
+
+void pulseaudio_lock()
+{
+	pa_threaded_mainloop_lock(pulseaudio_mainloop);
+}
+
+void pulseaudio_unlock()
+{
+	pa_threaded_mainloop_unlock(pulseaudio_mainloop);
+}
+
+void pulseaudio_wait()
+{
+	pa_threaded_mainloop_wait(pulseaudio_mainloop);
+}
+
+void pulseaudio_signal(int wait_for_accept)
+{
+	pa_threaded_mainloop_signal(pulseaudio_mainloop, wait_for_accept);
+}
+
+void pulseaudio_accept()
+{
+	pa_threaded_mainloop_accept(pulseaudio_mainloop);
+}
+
+int_fast32_t
+pulseaudio_get_source_info_list(pa_source_info_cb_t cb, void *userdata)
+{
+	if (pulseaudio_context_ready() < 0)
+		return -1;
+
+	pulseaudio_lock();
+
+	pa_operation *op = pa_context_get_source_info_list(
+		pulseaudio_context, cb, userdata);
+	if (!op) {
+		pulseaudio_unlock();
+		return -1;
+	}
+	while (pa_operation_get_state(op) == PA_OPERATION_RUNNING)
+		pulseaudio_wait();
+	pa_operation_unref(op);
+
+	pulseaudio_unlock();
+
+	return 0;
+}
+
+int_fast32_t
+pulseaudio_get_source_info(pa_source_info_cb_t cb, const char *name,
```

Comment:
Style: Don't line break between the function return type and the function name in a function declaration.